### DISCUSSION TITLE
Table columns don't resize with content

### DISCUSSION THREAD

**Discussion by weberc2** [fyne-io/fyne#3592] [General]
I followed https://developer.fyne.io/widget/table, but it seems like the column widths are set by the `"wide content"` string, and if I add a cell value that is wider than that initial string, the column won't expand (instead the label text just runs over onto adjacent cells). I would expect that adding values to a column that are wider than the initial column width would cause the column to resize dynamically (or perhaps wrap). Similarly, I would expect to be able to have columns with different widths depending on the longest value in each column. Is this possible with the existing TableWidget?

Here's a modified version of the widget/table example to illustrate:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/3314780/213878181-54b6d474-b88f-47bb-8872-7a20908ee7c7.png">

```go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/widget"
)

var data = [][]string{[]string{"really really really big value", "top right"},
	[]string{"bottom left", "really really really big value"}}

func main() {
	myApp := app.New()
	myWindow := myApp.NewWindow("Table Widget")

	list := widget.NewTable(
		func() (int, int) {
			return len(data), len(data[0])
		},
		func() fyne.CanvasObject {
			return widget.NewLabel("wide content")
		},
		func(i widget.TableCellID, o fyne.CanvasObject) {
			o.(*widget.Label).SetText(data[i.Row][i.Col])
		})

	myWindow.SetContent(list)
	myWindow.ShowAndRun()
}
```

**Comment by andydotxyz** [comment]
This is correct. The page you linked to specifies "... MinSize will define the standard size of each cell, and the minimum size of the table (it shows at least one cell)."

You probably want `SetColumnWidth`. This is not done automatically because the table widget is designed to handle huge content and negotiating a size based on the content of all cells is too slow.

**Reply by GaetanBaert** [reply]
Hello, 

I know this is a pretty old topic, but I want my columns to fit with the size of my largest string inside my table_data column when I update the data. I tried to find a correspondance factor between the value to give to `SetColumnWidth` and the length of the string, but I didn't succeed to find a good factor (mainly because the police is not monospace). I suppose I should use the `widget.Label.Size().Width` value, but I don't see how to do that, as I can't call `table.SetColumnWidth()` inside the `update` function when I create a table. Also, I can't check the current width of the column, so it is hard to know if my new value is greater than the current width. Do you have any tips for that ?

**Reply by andydotxyz** [reply]
> as I can't call table.SetColumnWidth() inside the update function

Why can't you?

>  I suppose I should use the widget.Label.Size().Width value

Actually the `Label.MinSize().Width` is the width of the column you want.

**Comment by Matt57** [comment]
That's too bad. It sounds like Fyne is too slow to implement a useful table. There is no reason to have a table that shows massive amounts of data if it cannot be read by the user. The way I see it, this table can only be useful if the data in the rows and columns are predictable and the width and height of each row and column are known ahead of time.

**Reply by andydotxyz** [reply]
I really don't know how you get "too slow" from the description - it is literally there to be really fast.

Content does not have to have a known size ahead of time - you can call SetColumnWidth and SetRowHeight when you load the content on each cell so the table adapts.

I'm sure you will understand that reading all the content from a table to resize each cell before we display anything is not a desirable thing to do, especially if you have a large or slow data source.

**Reply by andydotxyz** [reply]
All collection widgets (List, Tree, Table) laze load data, that is their design and it leads to excellent performance. If the concept is unfamiliar you can read more at https://developer.fyne.io/collection/list.
